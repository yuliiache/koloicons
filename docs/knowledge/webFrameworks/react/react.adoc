== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







