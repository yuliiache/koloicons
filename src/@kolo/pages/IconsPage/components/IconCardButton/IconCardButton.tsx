import ButtonIcon from '@kolo/uiKit/ButtonIcon/ButtonIcon';
import {ButtonIconType} from '@kolo/uiKit/ButtonIcon/constants';
import Icon from '@kolo/uiKit/Icons';
import {FC} from 'react';

import useIconCard from '../IconCard/hooks/useIconCard';

interface Props {
  iconId?: string;
}

const IconCardButton: FC<Props> = ({iconId}) => {
  const {isAdded, handleClick} = useIconCard();
  const actionIcon = isAdded ? <Icon.Close /> : <Icon.Add />;
  const tooltipLabel = isAdded ? 'Remove from collection' : 'Add to collection';

  return (
    <ButtonIcon
      id={iconId}
      type={ButtonIconType.ROUND}
      tooltipLabel={tooltipLabel}
      onClick={handleClick}
    >
      {actionIcon}
    </ButtonIcon>
  );
};

export default IconCardButton;
