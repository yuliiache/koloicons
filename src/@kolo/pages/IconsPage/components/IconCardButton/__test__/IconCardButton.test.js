import React from 'react';
import {render} from '@testing-library/react';

import IconCardButton from '../IconCardButton';

describe('IconCardButton', () => {
  it('renders correctly', () => {
    const {container} = render(<IconCardButton />);
    expect(container).toMatchSnapshot();
  });
});
