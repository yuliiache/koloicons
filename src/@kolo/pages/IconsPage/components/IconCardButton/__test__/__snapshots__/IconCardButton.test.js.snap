// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconCardButton renders correctly 1`] = `
<div>
  <div
    class="tooltipWrapper"
    data-testid="tooltipTestId"
    id="tooltip-Add to collection"
  >
    <button
      class="button button-round button-big"
      data-testid="buttonIconComponent"
    >
      <svg
        data-testid="icon-add"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8,2 C8.55228475,2 9,2.44771525 9,3 L9,3 L9,7 L13,7 C13.5128358,7 13.9355072,7.38604019 13.9932723,7.88337887 L14,8 C14,8.55228475 13.5522847,9 13,9 L13,9 L9,9 L9,13 C9,13.5128358 8.61395981,13.9355072 8.11662113,13.9932723 L8,14 C7.44771525,14 7,13.5522847 7,13 L7,13 L7,9 L3,9 C2.48716416,9 2.06449284,8.61395981 2.00672773,8.11662113 L2,8 C2,7.44771525 2.44771525,7 3,7 L3,7 L7,7 L7,3 C7,2.48716416 7.38604019,2.06449284 7.88337887,2.00672773 Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
